src/main/java/uk/gov/pay/connector/gateway/worldpay/WorldpayOrderBuilder.java

```
@@ -60,4 +63,21 @@ static GatewayOrder buildAuthoriseOrderWithExemptionEngine(CardAuthorisationGate
             return buildAuthoriseOrderWithoutExemptionEngine(aWorldpayAuthoriseOrderRequestBuilder(), request, acceptLanguageHeaderParser).build();
         }
     }
+
+    static GatewayOrder buildAuthoriseRecurringOrder(CardAuthorisationGatewayRequest request, AcceptLanguageHeaderParser acceptLanguageHeaderParser) {
+        WorldpayOrderRequestBuilder builder = (WorldpayOrderRequestBuilder) aWorldpayAuthoriseRecurringOrderRequestBuilder()
+                .withAgreementId(request.getAgreementId())
+                .withPaymentTokenId(request.getPaymentInstrument().map(paymentInstrument -> paymentInstrument.getRecurringAuthToken().get(WORLDPAY_RECURRING_AUTH_TOKEN_PAYMENT_TOKEN_ID_KEY)).orElse(""))
+                .withMerchantCode(AuthUtil.getWorldpayMerchantCode(request.getGatewayCredentials(), request.getAuthorisationMode()))
+                .withAmount(request.getAmount())
+                .withTransactionId(request.getTransactionId().orElse(""))
+                .withDescription(request.getDescription());
+
+        if (request.getPaymentInstrument().isPresent() &&
```

Could we just throw an exception if the payment instrument isn't present and get this at the top of the method? It would mean we wouldn't have to do multiple checks that it's present?

```
var paymentInstrument = request.getPaymentInstrument().orElseThrow(() -> new RuntimeException("Payment instrument not provided when trying to authorise a recurring payment")
```

---

👍